dags/nyc_taxi_dag.py

```diff
@@ -0,0 +1,139 @@
+            task_id="copy_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql="""
+                    COPY nyc_taxi.load_trips_staging
+                    FROM '{url}'
+                    WITH (format = 'csv', empty_string_as_null = true)
+                    RETURN SUMMARY;
+                """.format(
+                url=missing_url
+                )
+        ).execute(dict())
+
+        PostgresOperator(
+            dag=dag,
+            task_id="log_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql=Path('include/taxi-insert.sql').read_text(),
+        ).execute(dict())
+
+        PostgresOperator(
+            dag=dag,
+            task_id="mark_processed_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql="INSERT INTO nyc_taxi.load_files_processed VALUES ('{file}');".format(
+                file=missing_url
+                )
+        ).execute(dict())
+
+        PostgresOperator(
+            task_id="purge_staging_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql="DELETE FROM nyc_taxi.load_trips_staging;"
+        ).execute(dict())
+
+
+with DAG(
+    dag_id="nyc-taxi",
+    start_date=datetime.datetime(2021, 11, 11),
+    schedule_interval="@daily",
+    catchup=False,
+) as dag:
+    download_data_urls = SimpleHttpOperator(
+        task_id='download_data_urls',
+        method='GET',
+        http_conn_id='http_raw_github',
+        endpoint='toddwschneider/nyc-taxi-data/master/setup_files/raw_data_urls.txt',
+        headers={},
+        dag=dag
+    )
+
+    clean_data_urls = PythonOperator(
+        task_id="clean_data_urls",
+        python_callable=clean_data_urls,
+        provide_context=True,
+        op_kwargs={},
+    )
+
+    get_processed_files = PythonOperator(
+        task_id="get_processed_files",
+        python_callable=get_processed_files,
+        provide_context=True,
+        op_kwargs={},
+    )
+
+    identitfy_missing_urls = PythonOperator(
+        task_id="identitfy_missing_urls",
+        python_callable=identitfy_missing_urls,
+        provide_context=True,
+        op_kwargs={},
+    )
+
+    clean_data_urls << download_data_urls
+    identitfy_missing_urls << [clean_data_urls, get_processed_files]
+
+    process_new_files = PythonOperator(
+        task_id="process_new_files",
+        python_callable=process_new_files,
+        provide_context=True,
+        op_kwargs={},
+    )
+
+    process_new_files << identitfy_missing_urls
+
+    # The staging table should be empty already. Purging it again in case of
+    # an abort or other error case.
+    purge_staging_init = PostgresOperator(
+        task_id="purge_staging_init",
+        postgres_conn_id="cratedb_demo_connection",
+        sql="DELETE FROM nyc_taxi.load_trips_staging;"
+    )
+
+    process_new_files << purge_staging_init
```

Comment:
Maybe it would be easier to parse when all `Operator` are at the beginning of the block and building of DAG is blocked.

I think I would then also use same syntax as for `identitfy_missing_urls`: 

```
process_new_files << [purge_staging_init,  purge_staging_init]
```

Reply:
Rearranged the dependency declarations a bit so that they are made right after the corresponding operator. That also allowed merging the condition for `process_new_files`,  as you suggested.